### ISSUE TITLE
Browserstack End to End Tests failing

### ISSUE THREAD

**Issue by jgadsden** [OWASP/threat-dragon#402]
**Describe what problem your feature request solves**
githcub  action 'Browserstack End to End Tests' are failing

**Describe the solution you'd like**
Not sue why, but it seems to be the e2e-nightly fails some of its tests

**Additional context**

**Comment by lreading** [comment]
Looking at the browserstack logs, it seems it's having trouble finding the v2 demo threat model in some cases.  This is the report spec, and I saw similar behavior in a recent PR and had to re-run the tests.  I'll see if I can find a way of making it less flaky, though I've no idea why it isn't working.  From the logs:
> Timed out retrying after 4000ms: Expected to find content: 'Version 2 Demo Model' within the elemen

Screenshot:
![Screenshot from 2022-04-19 21-46-25](https://user-images.githubusercontent.com/5851985/164130353-bcfaab4d-a649-4d6a-a9d1-3b90554da03c.png)

**Comment by lreading** [comment]
The response times for the web application are terrible. From BrowserStack, it can take about 10 seconds to load the main ThreatDragon page. Looking at the Heroku dashboard, it seems like response times can be as slow as 600ms, however, we aren't anywhere near max capacity for our plan. Something funny is happening here at the infrastructure level (either BrowserStack or Heroku).  Not entirely sure which one yet.  I am testing if adding a wait between some of the calls will make any difference.

**Comment by lreading** [comment]
Firefox is still struggling with the report spec even after adding the waits.  There are a bunch of tests that are failing to launch in BrowserStack, and I couldn't find any additional information on the error.
![Screenshot from 2022-04-19 22-25-57](https://user-images.githubusercontent.com/5851985/164134767-608199bf-436e-45cb-9e4c-cdda69c67241.png)

When you hover over it, it says "Spec details unavailable for this combination as it was terminated."
Additionally, I'm unable to even load the BrowserStack dashboard on Firefox (linux)... I had to use Chrome.  Very odd behavior.

I opened a support ticket with BrowserStack regarding these tests failing to run.  Depending on what we hear from BrowserStack, we may consider another cross-browser testing option that supports cypress?  BrowserStack has been very generous as to provide us with a free plan as we're an open-source project, but it wouldn't hurt to explore other options as well.

**Comment by lreading** [comment]
Apologies, I did not intend to close this issue.  Reopened, waiting on a response from Browserstack to continue debugging.

**Comment by jgadsden** [comment]
OK, thanks @lreading  - I have some ideas on how to (possibly) fix / cope with this ... all to do with what Browserstack regards as a failure ontheir side (which does not fail our run) and a failure our side (which does fail our run)

Do you want me to try them out?
* we make sure each test is self contained - so no dependency on a previous test. Browserstack flattens the directory structure and then runs the specs in alphabetical order
* I had this issue when running Cypress e2e smoke tests locally, and bodged a fix by forcing the smoke tests to run in a defined order. I am not proud of that bodge, and am looking to do it correctly

**Comment by jgadsden** [comment]
OK, [this browserstack run](https://automate.browserstack.com/dashboard/v2/builds/cbfff9c379a43823d39e3b2e6c65eea15c9e2bb7) shows that threatdragon/report.js spec is dependent on ordering/session/something - I just removed it and all is passing (except for some of the BrowserStack machines not being available).

We should probably look at all the tests and ensure that they leave the state as they found it.

**Comment by lreading** [comment]
I'm not sure that I'm seeing that in the run you referenced.  What I'm seeing (by looking at the playback videos) is that it's starting from a fresh state and not finding/clicking the version 2 demo model for some reason.  My understanding of how Cypress works is that each spec will be treated as its own isolated test case, with the `before` section running similar to how a before block behaves in a karma test.  I think that by starting each test suite from the home page and navigating to the demo model, it should be clearing the state and resetting any changes. (triggering this flow: https://github.com/OWASP/threat-dragon/blob/main/td.vue/src/store/modules/threatmodel.js#L64 )... But maybe not?

I'm not saying there's nothing funky happening with state/order here (and if there is, we most definitely want to fix it!).  I'm going to run the smokes locally a bunch of times and see if I can reproduce what you were seeing.  What I'm seeing looks more like a selector issue to me, but I don't know if I'm looking at it the same way or in the same places as you.   :smile:

**Comment by jgadsden** [comment]
I will have a look at report.js later today and see if we can get it to pass the browserstack tests

**Comment by lreading** [comment]
I'm also chipping away at it.  I think I found a reasonable solution to ensuring the state is cleared between tests.  In my testing branch, I've added a parent command that should be called in the before section for each test, `cy.setupTest()`

This command visits the home page, clears the sessionStorage, and also closes the v2 notification toast.  When we get rid of that toast, we will only need to update it in one spot.

```
    cy.visit('/', {
        onBeforeLoad: (win) => win.sessionStorage.clear()
    });
    cy.get('.Vue-Toastification__close-button').click();
```

**Comment by lreading** [comment]
I was able to get a clean run.  It seems as though cypress didn't like selecting an anchor by text.  I've added a data attribute and am selecting the link by that, and it seems to work as expected.  Some tests are still not launching, and I will take that back up when Browserstack has a chance to get back to me.  That part seems to be an issue on their end, not ours. :smile:

https://github.com/OWASP/threat-dragon/runs/6113245924

I did find the smoke test you were referring to, and the "order" issue is now resolved by the "setupTest".  I'll have a PR soon!
